### PR TITLE
parallel_for_2d fix and guarding avx/avx2 compilation

### PR DESCRIPTION
**Pull request by cpuhrsch**
Fix for https://github.com/pytorch/pytorch/issues/5921.

I'm adding support compilers that don't support -mavx -mavx2 by revisiting the dispatch code. This is work in progress, but I'll attempt to make it part of this PR.

I'm waiting for my remote devserver to come online again so that I can run performance regression tests. I'm currently developing on my Macbook.

### CODE REVIEW

**Comment by cpuhrsch** [comment]
On my machine the tests mentioned in https://github.com/uber/pyro/issues/912#issuecomment-375037044 run clear now

In particular

```
(base) [14:28:59: cpuhrsch@devfair0129 pyro]$ pytest -v --tb=short tests/infer/test_enum.py::test_elbo_iarange_iarange[2-2-None-None-parallel-None]
============================================================= test session starts =============================================================
platform linux2 -- Python 2.7.14, pytest-3.4.2, py-1.5.2, pluggy-0.6.0 -- /private/home/cpuhrsch/miniconda2/bin/python
cachedir: .pytest_cache
rootdir: /private/home/cpuhrsch/dev/pyro, inifile: setup.cfg
collected 1 item

tests/infer/test_enum.py::test_elbo_iarange_iarange[2-2-None-None-parallel-None] PASSED                                                 [100%]

========================================================== 1 passed in 0.48 seconds ===========================================================
(base) [14:29:10: cpuhrsch@devfair0129 pyro]$
```

File: test/test_torch.py [hunk 1]
```diff
@@ -542,6 +542,37 @@ def test_multidim(x, dim):
     def test_dim_reduction(self):
         self._test_dim_reduction(self, lambda t: t)
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_cpu_parallel(self):
+        # To use parallel branches we'll need to compare on tensors
+        # that are relatively large. Even if this is run on a single
+        # core machine these tests will still give you signal on
+        # the correctness
+        pass
```
Comment (soumith) [inline test/test_torch.py]: forgot to remove?

**Comment by neerajprad** [comment]
@cpuhrsch - Thanks! I confirmed too that this passes all our tests.

**Review by soumith — APPROVED** [review]

**Comment by cpuhrsch** [comment]
Hold off on landing this just yet, I want to add a few more lines.

**Comment by cpuhrsch** [comment]
Once all tests pass this is good to go from my end.

File: test/test_torch.py [hunk 2]
```diff
@@ -542,6 +542,36 @@ def test_multidim(x, dim):
...
+                    tvs = tv.sum(dim)
+                diff = np.abs(nvs - tvs.numpy()).sum()
+                self.assertEqual(diff, 0)
+
+        sizes = []
+        sizes += [[2, 3, 3, 3, 3, 2, 2, 3, 2, 3, 2, 3, 3]]
+        sizes += [[4, 4, 4, 4, 4, 4, 4, 4, 4, 4]]
+        sizes += [[1, 32 * 8 * 32 * 8]]
+        sizes += [[1, 32770]]
+        map(_run_test, sizes)
```
Comment (apaszke) [inline test/test_torch.py]: map is lazy in Python3 so this won’t run even a single test :confused:

File: test/test_torch.py [hunk 3]
```diff
@@ -542,6 +542,36 @@ def test_multidim(x, dim):
...
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_cpu_parallel(self):
+        # To use parallel branches we'll need to compare on tensors
+        # that are relatively large. Even if this is run on a single
+        # core machine these tests will still give you signal on
+        # the correctness
+
+        def _run_test(size):
+            for dim in range(len(size) + 1):
+                nv = np.round(np.random.rand(*size))  # 0s and 1s
```
Comment (apaszke) [inline test/test_torch.py]: Can’t we just use .bernoulli_() and drop the numpy dependency?

Comment (cpuhrsch) [reply test/test_torch.py]: No, since we're using numpy's sum as well to compare with torch's sum.

File: test/test_torch.py (hunk 2 above)
Comment (cpuhrsch) [reply test/test_torch.py]: PR is [here](https://github.com/pytorch/pytorch/pull/5940)

File: test/test_torch.py (hunk 3 above)
Comment (apaszke) [reply test/test_torch.py]: Couldn't we use `.tolist()`, flatten the lists and use the `sum()` Python builtin?

**Pull request merged**